interlocking/interlockingcontroller/trackcontroller.py

```diff
@@ -13,107 +15,111 @@ def __init__(self, interlocking, point_controller, signal_controller):
+        segments_of_route_2 = set(map(lambda seg: seg.segment_id, route_2.get_segments_of_route))
         if len(segments_of_route_1.intersection(segments_of_route_2)) > 0:
             return True
 
         overlaps_of_route_1 = route_1.get_overlaps_of_route()
         found_no_free_overlap_route_1 = True
         for overlap in overlaps_of_route_1:
-            overlap_segments_of_route_1 = {x for v in overlap.segments.values() for x in v}
-            if len(overlap_segments_of_route_1.intersection(segments_of_route_2)) == 0:
+            overlap_segments = set(map(lambda seg: seg.segment_id, overlap.segments))
+            if len(overlap_segments.intersection(segments_of_route_2)) == 0:
                 found_no_free_overlap_route_1 = False
                 break
         overlaps_of_route_2 = route_2.get_overlaps_of_route()
         found_no_free_overlap_route_2 = True
         for overlap in overlaps_of_route_2:
-            overlap_segments_of_route_2 = {x for v in overlap.segments.values() for x in v}
-            if len(overlap_segments_of_route_2.intersection(segments_of_route_1)) == 0:
+            overlap_segments = set(map(lambda seg: seg.segment_id, overlap.segments))
+            if len(overlap_segments.intersection(segments_of_route_1)) == 0:
                 found_no_free_overlap_route_2 = False
                 break
         return found_no_free_overlap_route_1 or found_no_free_overlap_route_2
 
-    def free_route(self, route):
-        self.overlap_controller.free_overlap_of_route(route)
-
-    def reset_route(self, route):
-        segments = route.get_segments_of_route()
-        for track_base_id in segments:
-            track = self.tracks[track_base_id]
-            for segment_id in segments[track_base_id]:
-                print(f"--- Set track {segment_id} free")
-                track.state[segment_id] = "free"
-
-        self.overlap_controller.free_overlap_of_route(route)
-
-    def reserve_route(self, route):
-        segments = route.get_segments_of_route()
-        for track_base_id in segments:
-            track = self.tracks[track_base_id]
-            for segment_id in segments[track_base_id]:
-                print(f"--- Set track {segment_id} reserved")
-                track.state[segment_id] = "reserved"
-        self.overlap_controller.reserve_overlap_of_route(route)
-
-    def occupy_segment_of_track(self, track, segment_id):
-        if track.state[segment_id] != "occupied":
-            print(f"--- Set track {segment_id} occupied")
-            track.state[segment_id] = "occupied"
+    def free_route(self, route, train_id: str):
+        for segment in route.get_segments_of_route():
+            if segment.state != OccupancyState.FREE:
+                print(f"--- Set track {segment.segment_id} free")
+                segment.state = OccupancyState.FREE
+                segment.used_by.remove(train_id)
```

Comment:
I am not sure, if it is helpful to free the segments in free_route. Previously the segments were only freed from `free_segment_of_track` and I think that is better. Any good tds should free every segment individualy. If you free the route befor freeing the segment, that will do nothing anymore. Maybe here it should instead check, if every part of the route (segments and maybe points) have been freed alredy. 
This check could also be done in the interlocking.

What do you think?

Reply:
If you change that, reset route will have to change as well.

Reply:
Hmm... I'm pretty sure I had a very good reason for this, since I had the same thought in mind. But I can't remember that reason. So I changed the code as you suggested: The `free_route` method checks and prints an error (maybe it should also have consequences) and the `reset_route` resets all segments to free.